KAFKA-9374: Make connector interactions asynchronous

C0urante opened the pull request:
[Jira ticket](https://issues.apache.org/jira/browse/KAFKA-9374)

These changes allow herders to continue to function even when a connector they are running hangs in its start, stop, initialize, validate, and/or config methods.

The main idea is to make these connector interactions asynchronous and accept a callback that can be invoked upon the completion (successful or otherwise) of these interactions. The distributed herder handles any follow-up logic by adding a new herder request to its queue in that callback, which helps preserve some synchronization and ordering guarantees provided by the current tick model.

If any connector refuses to shut down within a graceful timeout period, the framework will abandon it and potentially start a new connector in its place (in cases such as connector restart or reconfiguration).

Existing unit tests for the distributed herder and worker have been modified to reflect these changes, and a new integration test named `BlockingConnectorTest` has been added to ensure that they work in practice.

C0urante committed:
KAFKA-9374: Make connector interactions asynchronous

These changes allow herders to continue to function even when a connector they
are running hangs in its start, stop, initialize, validate, and/or config
methods.

ncliang commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -270,6 +273,8 @@ public boolean startConnector(
             if (existing != null)
                 throw new ConnectException("Connector with name " + connName + " already exists");
 
+            executor.submit(workerConnector);
```
It looks like you are reusing the executor that currently is used to run the `WorkerTask`. Since operations are async now, would it be possible to have `startTask` called on a different thread while connector is still being initialized or is transitioning to started state here?

C0urante replied:
As far as the API provided by the `Worker` class goes, yes, it is possible and there are no guards in place to prevent that from happening. However, at the moment both herders are well-behaved in that they do not instruct the `Worker` to start tasks until after the connector has finished starting as well. This is pretty simple to verify, and unlikely to change, since the herders wait for connectors to start before requesting task configurations from them, which in turn are required in order to actually go about starting tasks.

ncliang commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorPluginsResource.java:
```diff
@@ -78,7 +79,11 @@ public ConfigInfos validateConfigs(
+        herder.validateConnectorConfig(connectorConfig, validationCallback);
+
+        // TODO: Timeout?
+        return validationCallback.get();
```
Probably a good idea to have timeout to not block indefinitely.

C0urante replied:
Ack, addressed

ncliang reviewed (commented):
@C0urante Thanks for making this change. It is not an easy change to review. But, in general, it seems complex but may be necessary to support pre-emption use case. I probably would have opted for a simpler solution to just allow transition to STOP state to preempt instead of the general case solution you have here. I just have a couple comments and questions.

C0urante commented:
There were a few questions I tried to answer before pushing these changes, specifically with regards to the distributed herder. Proactively publishing my findings here to (hopefully) help provide confidence that nothing should break as a result of these.


# Is it safe to split some behavior across two or more herder requests (between which a rebalance may have occurred)?

This applies to these cases:

## Writing a new connector config

The first request begins validating the configuration, which is handled asynchronously. Once validation is completed, a callback then adds a second external request to the queue that either forwards the configs to the leader or, if the worker is the leader, writes them to the config backing store.

The only relevant impacts that a rebalance in between these two requests may have is the election of a new leader and/or the updating of the config snapshot.

The election of a new leader should be handled cleanly as that information is both accessible and stable during the second request, and not utilized during the first.

The updating of the config snapshot should have no adverse impact on the second request. If the connector is being created via the `POST /connectors` endpoint (which fails if the connector already exists in the config backing store), then it is possible this request will fail during the second request if the updates to the snapshot between the two requests contain a configuration for the to-be-created connector. However, this will only occur if there are concurrent attempts to create the connector, which already poses the easily-predicted risk that one will fail.

Additionally, with reconfiguration requests (i.e., requests to the `PUT /connectors/{connector}/config` endpoint) it's possible that another connector configuration may be submitted before the validation of the first configuration completes. If the validation for the first configuration takes longer than the validation for the second, the second configuration may be written to the config backing store first, and then overwritten by the first configuration once validation for it completes. This should not be a common case as connector creation should already be performed with care by users of the framework. However, if necessary, we could consider adding pre-emption logic for connector reconfiguration requests that ensures that the last request always wins.

## Restarting a connector

The first request stops the connector (which no longer blocks on the call to `Connector::stop`, but does schedule the connector to be shut down as soon as it returns control to the framework), and then begins asynchronously starting the connector. Once the connector has finished starting (meaning it has been initialized and transitioned to its initial target state), a callback then adds a second external request to the queue that, if the connector has started running (as opposed to being created in the `PAUSED` target state), runs the `reconfigureConnectorTasksWithRetry` method. This method is responsible for (with retry logic) requesting task configs from the connector and, if they differ from the existing task configs for the connector, either writing them to the config backing store (if the worker is the leader) or forwarding them to the leader (if the worker is not the leader). The `reconfigureConnectorTasksWithRetry` method is effectively a no-op if the connector is not currently owned and being run by the worker.

**There is currently no guard in place to prevent zombie connector threads from invoking that method.** There probably should be. Address in a follow-up PR?

There is a guard in the `reconfigureConnectorTasksWithRetry` method that short-circuits the method and essentially turns it into a no-op if the to-be-reconfigured connector is not currently running on the worker. This, in combination with the check before writing to the task configs topic (which should render the method idempotent as long as the task configs for the connector do not change), should make it safe to invoke the `reconfigureConnectorTasksWithRetry` method in a separate herder request, even if the worker's assignment or the leader of the cluster change in between the first and second herder requests.

## Resuming a connector

Pausing a connector is performed asynchronously, but does not require a new herder request. However, the framework currently handles resumption of a connector by immediately requesting task configs after the connector has been started and, if necessary, writing them to the config backing store. This action (requesting these configs and possibly writing them to the backing store) is now accomplished via a herder request, which is added to the request queue via callback if and when the connector is started successfully.

For the same reasons as with the case of restarting a connector, this should be safe.


# Is it safe to perform some (or all) of this behavior, which used to be performed during a different part of the tick cycle, during the "external requests" portion of the tick cycle now?

## Writing a new connector config

All of this takes place through a herder request at the moment, so this is fine.

## Restarting a connector

All of this takes place through a herder request at the moment, so this is fine.

## Resuming a connector

The logic of communicating the state change to the worker is accomplished synchronously in the same location as it currently is (the `processTargetStateChanges` method of the herder's tick cycle). The only logic that has been moved outside of this method is the generation of task configs for the connector, which is accomplished via herder request under other circumstances (specifically, when a connector invokes `ConnectorContext::requestTaskReconfiguration`).

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:
```diff
@@ -299,103 +304,115 @@ public StatusBackingStore statusBackingStore() {
-                                                               ConnectorClientConfigRequest.ClientType.ADMIN,
-                                                               connectorClientConfigOverridePolicy);
+    public void validateConnectorConfig(Map<String, String> connectorProps, Callback<ConfigInfos> callback) {
+        connectorExecutor.submit(new Runnable() {
```
It'd be great to minimize the number of changes lines, so maybe put the runnable logic as a separate method to eliminate the indentation-only changes.

C0urante replied:
Ack, addressed within the `AbstractHerder` class and will see if similar changes can be made elsewhere.

C0urante commented:
@ncliang I've made some updates to the PR and rebased on the latest trunk; would you be willing to do another pass?

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -263,17 +267,20 @@ public boolean startConnector(
+                onConnectorStateChange.onCompletion(t, null);
+                return;
             }
+            workerConnector.transitionTo(initialState, onConnectorStateChange);
```
This part still needs some work; it's in an inconsistent state because I modified `Worker::startConnector` to have no return value and instead communicate all success or failure of the connector startup through the callback, but haven't taken care of issues like possibly invoking the callback twice (once in this method, and once in the `WorkerConnector` instance), making sure to swap plugin classloaders at the right times, and preventing a possible race with the check to see if the connector already exists based on whether its name is present as a key in the `connectors` map.

C0urante replied:
No race is possible; `connectors` is a `ConcurrentHashMap` and its `putIfAbsent` method is atomic.

We can just move this line (`workerConnector.transitionTo(initialState, onConnectorStateChange);`) back into the `try` block; it may throw an exception, but the callback we pass in will never be invoked until the `workerConnector` instance is submitted to the executor, so there should be no risk of it being invoked twice.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/HerderConnectorContext.java:
```diff
@@ -16,30 +16,64 @@
 
     @Override
     public void requestTaskReconfiguration() {
+        synchronized (this) {
```
why use locking if the variable is `volatile` and can change only to `true` during the lifetime of this object?

C0urante replied:
Ah, good point. No need, will remove the locking.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -266,31 +422,51 @@ public void close() {
         public void onStartup(String connector) {
             state = AbstractStatus.State.RUNNING;
-            delegate.onStartup(connector);
+            synchronized (this) {
```
same question here and below about locking around a `volatile` variable. Is this the only reason to lock here? One would think so based on previous usage.

C0urante replied:
The locking here is to ensure that, upon being cancelled, the very last status update that this connector makes is to set its state to `UNASSIGNED`.

There's a potential race if its status is scheduled to be updated to, e.g., `PAUSED` and the check for `cancelled` goes through because it isn't set to `true` yet, then the task gets cancelled on another thread and its status gets set to `UNASSIGNED`, then the original thread proceeds with execution and sets the status to `PAUSED`.

kkonstantine replied:
I don't think locking buys you anything w/r/t to that. The producer#send in the status backing store is asynchronous. So what would describe above can happen anyways, regardless of whether you lock this object. 

Of course, if it wasn't asynchronous things would be much worse. A bottleneck would be created by the lock, waiting for the `send` to finish in every locked block, so that's not an option. Wdyt? 
That's what I see at the high level without spending to much time on it, but see if you can check this assumption and we can return to this question.

C0urante replied:
AFAICT the underlying `KafkaBasedLog` used for the status store does provide ordering guarantees: https://github.com/apache/kafka/blob/1d438033f7f4d7cbdac34e1b57d97b5a647ae4de/connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java#L242-L249

kkonstantine replied:
Thanks for checking the underlying implementation @C0urante . 

That takes us to my earlier concern about this operation potentially blocking for too long to be in a `synchronized` block. And the potential of blocking does not have to do with acknowledging that the record was written only. The producer call has a metadata update call too. 

Going over the uses of `KafkaBasedLog` in Connect, I didn't find an example where we have `KafkaBasedLog#send` running  in mutual exclusion. Contrary, similar concerns are probably the reason why we call `OffsetStorageWriter#doFlush` outside the synchronized block in `WorkerSourceTask`. 

I think we might be able to live with a rare race condition as the one you described, in order to avoid introducing unintended side-effects due to locking.

C0urante replied:
I'd rather not knowingly allow a race here if possible, and I think it should be fine to synchronize here, but if we do go that route I can file a pre-emptive jira ticket and link to it in the source code so that anyone who does run into it has at least some chance of understanding what's going on (and possibly refactoring the framework to improve things).

The framework does synchronize around all status updates for tasks, as you can see in the `WorkerTask` class:

`TaskStatus.Listener::onStartup`: https://github.com/apache/kafka/blob/0f68dc7a640b26a8edea154ea4ea2b6d93b5104b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L172-L182

`TaskStatus.Listener::onShutdown`: https://github.com/apache/kafka/blob/0f68dc7a640b26a8edea154ea4ea2b6d93b5104b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L195-L202

`TaskStatus.Listener::onFailure`: https://github.com/apache/kafka/blob/0f68dc7a640b26a8edea154ea4ea2b6d93b5104b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L206-L213

`TaskStatus.Listener::onPause`: https://github.com/apache/kafka/blob/0f68dc7a640b26a8edea154ea4ea2b6d93b5104b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L216-L218

`TaskStatus.Listener::onResume`: https://github.com/apache/kafka/blob/0f68dc7a640b26a8edea154ea4ea2b6d93b5104b/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L220-L222

kkonstantine replied:
Yeah, good point. I've been on the fence here, because the calls might block one way or the other anyway. Good thing is that the locking is per connector instance. 

Let's keep the synchronization. We can always return with some performance benchmarking on the topic and revisit locking for connectors _and_ tasks if we can afford to do that.

But let's follow the `synchronized` method style (e.g. `onPause`, `onResume` in your examples above) if the synchronization applies to the whole method.

C0urante replied:
Sounds good, will address the synchronization style in the next commit.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -191,32 +192,71 @@ public synchronized void putConnectorConfig(String connName,
+                        return;
+                    }
+
+                    requestExecutorService.submit(new Runnable() {
```
I don't think this is worth backporting before AK 2.0, and maybe not even as far back. Given that, I'd suggest using lambda notation whenever a new Runnable is needed to avoid the Runnable declaration boilerplate.

C0urante replied:
👍 Ack, will change

C0urante replied:
In order to avoid mixing style, do you think it'd be best to replace _all_ instances of `Runnable` construction with lambda notation in files changed by this PR?

kkonstantine reviewed (commented):
Great to see activity on this one. 

Just dropping a few comments I noticed, but I haven't finished a full pass yet. Will return.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -246,15 +251,16 @@ public boolean startConnector(
             final WorkerConnector workerConnector;
             ClassLoader savedLoader = plugins.currentThreadLoader();
             try {
+                final String connClass = connProps.get(ConnectorConfig.CONNECTOR_CLASS_CONFIG);
```
:+1:

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -166,27 +244,105 @@ private void pause() {
             if (state == State.STARTED)
                 connector.stop();
-            this.state = State.STOPPED;
+            WorkerConnector.this.state = State.STOPPED;
```
nit: seems unnecessary?
similar for `State.FAILED` below.

C0urante replied:
Mmm, good catch. Think this might have been left behind from a previous iteration of the PR.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -1182,28 +1251,47 @@ public Void call() throws Exception {
-                    startConnector(connectorName);
+                    startConnector(connectorName, (error, result) -> { });
                 } catch (Throwable t) {
                     log.error("Couldn't instantiate connector " + connectorName + " because it has an invalid connector " +
```
Is it possible to trigger this log statement now?
Should this be moved into the callback?

C0urante replied:
Since we're catching `Throwable` here, yes, I do think it's still technically possible to trigger the log statement. But you're definitely right that the actual content of that message should be moved into the callback; thinking the log message here in the `catch` block can be made even scarier since if something goes wrong here that's really unexpected.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:
```diff
@@ -82,6 +82,9 @@
     // we need to consider all possible scenarios this could fail. It might be ok to fail with a timeout in rare cases,
     // but currently a worker simply leaving the group can take this long as well.
     public static final long REQUEST_TIMEOUT_MS = 90 * 1000;
+    // Mutable for integration testing; otherwise, some tests would take at least REQUEST_TIMEOUT_MS
```
If only time was mocked.

C0urante replied:
I mean, we have stuff to do that (the `Time` interface and the `MockTime` and `SystemTime` implementations of it) but it doesn't seem like it'd be easy to leverage here, since the request timeout is performed by a call to `FutureCallback::get`, which relies on `ShutdownLatch::await` under the hood.
AFAIK we don't have a great way to mock shutdown latch timeouts, but if we do, it might not be so bad to do that instead of this hack here.

gharris1727 replied:
Yeah, I don't think that it's worth changing everything just to mock time, especially if requires us to change the functionality.
SGTM.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:
```diff
@@ -97,6 +100,15 @@ public ConnectorsResource(Herder herder, WorkerConfig config) {
     }
 
+    // For testing purposes only
+    public static void setRequestTimeout(long requestTimeoutMs) {
```
Can this be knocked down to protected / package-private?

C0urante replied:
Don't think so :(
Right now it's really only useful to call this from an integration test, and those all live in a different package. We could move to a different package but it seems a little funky to put one of our integration tests in a different place from all the others, or to move all of our integration tests into the seemingly-arbitrary `org.apache.kafka.connect.runtime.rest.resources` package.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -191,32 +192,61 @@ public synchronized void putConnectorConfig(String connName,
-        } catch (ConnectException e) {
-            callback.onCompletion(e, null);
+                requestExecutorService.submit(() -> {
+                    synchronized (this) {
```
All of the call sites for this function have synchronization blocks, can you just add the `synchronized` keyword to the method instead?

C0urante replied:
Ah yeah, done. Could have sworn I was getting SpotBugs complaints at one point when I tried that, but it seems to work now.

gharris1727 reviewed (commented):
This looks great, and is an incredibly useful change!
I deleted a lot of comments after I realized what you were doing, and only found a few nits.

kkonstantine reviewed (commented):
Replied on the question around locking. 
Also, 
> In order to avoid mixing style, do you think it'd be best to replace all instances of Runnable construction with lambda notation in files changed by this PR?

I can't return to this question in place, but sure, if there's only a few other uses in relevant files we could unify the style.

C0urante commented:
Thanks for the review, @gharris1727. I've addressed all of your comments with either changes or responses; let me know what you think.

@kkonstantine -- there are a few places with nested callback logic that I've left untouched for now just to avoid inflating the diff, but apart from those, I've replaced construction of anonymous `Runnable` and `Callback` instances with lambda syntax in the `StandaloneHerder` and `DistributedHerder` classes. Tried to strike a balance here; if we want to go further we always can.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -239,9 +236,10 @@ private synchronized void putConnectorConfig(String connName,
+                    updateConnectorTasks(connName);
+                    // synchronized (this) {
+                    //     updateConnectorTasks(connName);
+                    // }
```
nit: leftover comments

C0urante replied:
🤦‍♀️ thanks

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -412,9 +407,10 @@ public void onConnectorTargetStateChange(String connector) {
+                            updateConnectorTasks(connector);
+                            // synchronized (StandaloneHerder.this) {
+                            //     updateConnectorTasks(connector);
+                            // }
```
nit: leftover comments

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -536,20 +561,37 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
             // The update may be a deletion, so verify we actually need to restart the connector
             if (remains)
-                startConnector(connectorName);
+                startConnector(connectorName, (error, result) -> { });
```
The caller of this function doesn't need to block on starting the connector, but I think it should log errors inside of the callback. Otherwise they're going to get swallowed.

C0urante replied:
Agh, you're right. Thought I'd committed that but apparently not. Will add.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -1182,31 +1245,49 @@ public Void call() throws Exception {
             public Void call() throws Exception {
                 try {
-                    startConnector(connectorName);
+                    startConnector(connectorName, (error, result) -> { });
```
This needs a log statement to avoid swallowing the exception silently.

gharris1727 reviewed (approved):
LGTM, thanks for this great improvement!

rhauch commented:
retest this please

rhauch commented:
This should be backported at least to the `2.6` branch.

rhauch commented:
retest this please

rhauch commented:
Green builds, except for JDK 11 which failed to run due to unrelated build issues.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/CloseableConnectorContext.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.io.Closeable;
+
+public interface CloseableConnectorContext extends ConnectorContext, Closeable {
```
Any reason not to use `AutoCloseable` instead?

C0urante replied:
It's not used in a try-with-resources block right now, and it's pretty unlikely that it will be. This also follows the precedent with the [CloseableOffsetStorageReader](https://github.com/apache/kafka/blob/0f68dc7a640b26a8edea154ea4ea2b6d93b5104b/connect/runtime/src/main/java/org/apache/kafka/connect/storage/CloseableOffsetStorageReader.java) class, which is also not `AutoCloseable`.

Not really attached either way, happy to change to `AutoCloseable` if there's a reason to.

kkonstantine replied:
It's more that `Closeable` is the legacy interface and we don't have strict requirements on throwing `IOException` or this call being idempotent.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -356,40 +390,96 @@ public boolean isSinkConnector(String connName) {
+                return;
             }
 
+            log.info("Stopping connector {}", connName);
```
not sure what's the advantage of bringing this log message further down here. Searching the logs for `Stopping connector` will miss the cases of "unowned" connectors, making debugging potentially more challenging.

C0urante replied:
My original thinking was that it doesn't make sense to log that we're stopping a connector, even if we log immediately afterward that we're not actually stopping that connector since we don't own it.

On second thought, probably best to keep it as-is just to avoid complicating things for people who are used to reading these log lines in succession. I'll revert this change.

kkonstantine replied:
if/else branching when parsing logs can be tricky. The log indicates that the worker is indeed in the process of stopping a connector that was requested. The next log message will indicate that it didn't actually own the requested connector.

kkonstantine reviewed (commented):
Thanks for addressing previous comments @C0urante 
I added a couple more, I'm half way in my review. 
Mainly I returned to the question around locking. I still think that locking around producer calls might be something we'd be better off avoiding. Wdyt?

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -239,38 +243,59 @@ public void stop() {
+                final Connector connector = plugins.newConnector(connClass);
+                final ConnectorConfig connConfig = ConnectUtils.isSinkConnector(connector)
+                    ? new SinkConnectorConfig(plugins, connProps)
+                    : new SourceConnectorConfig(plugins, connProps, config.topicCreationEnable());
```
I think I've noticed that you tab size when lines are wrapped is not what we've been using. As you see below, this indentation corresponds to two tabs (8 characters). Only exception I've found that is enforced by checkstyle is when a lambda is wrapped. But other than than we've been using 2 tabs for multiline statements.

C0urante replied:
Ack, will try to make the adjustment. May miss a few places just due to the size of the PR.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -239,38 +243,59 @@ public void stop() {
-                final ConnectorConfig connConfig = workerConnector.isSinkConnector()
-                        ? new SinkConnectorConfig(plugins, connProps)
-                        : new SourceConnectorConfig(plugins, connProps, config.topicCreationEnable());
+                    offsetBackingStore,
```
nit: keeping the previous format of grouping arguments in a few lines matches what we have in other versions. This change was brought in when the `OffsetStorageReader` PR was merged but from the diff you can see that it was changed back.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -165,27 +241,115 @@ private void pause() {
+     * #{@link #awaitShutdown} to block until completion.
+     */
     public void shutdown() {
+        synchronized (this) {
```
if you endup locking the whole method, there's no reason to use a block, you can use `synchronized` on the method. Applies here and any similar method.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -337,6 +340,12 @@ public void tick() {
         }
 
         // Process any external requests
+        // TODO: Some of these can be performed concurrently or even optimized away entirely.
```
I wonder if we should create a jira issue instead

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -414,31 +423,44 @@ public void tick() {
 
         // Let the group take any actions it needs to
         try {
+            log.trace(
```
nit: our style allows the first line to be with the log statement (see your other trace call below)

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -486,6 +509,10 @@ private synchronized boolean updateConfigsWithEager(AtomicReference<Set<String>>
                 }
             }
+        } else {
+            log.trace("Skipping config updates with eager rebalancing "
```
do we know how often this message or the one below would be printed?

C0urante replied:
Not sure how often but I can think of some scenarios that would cause it:

- Session key is updated
- The number of workers in the cluster changes

Why do you ask?

kkonstantine replied:
I'm trying to understand if it would overwhelm the logs at the TRACE level

C0urante replied:
Ahh, gotcha. It'll definitely make them a bit noisier, but this line will only ever be reached once per herder tick, which already produces several other messages at a higher level.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -540,20 +572,45 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
         // If we only have connector config updates, we can just bounce the updated connectors that are
         // currently assigned to this worker.
         Set<String> localConnectors = assignment == null ? Collections.<String>emptySet() : new HashSet<>(assignment.connectors());
+        log.trace(
```
nit; multiline splitting is a bit unusual

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -540,20 +572,45 @@ private void processConnectorConfigUpdates(Set<String> connectorConfigUpdates) {
-            if (remains)
-                startConnector(connectorName);
+            if (remains) {
+                startConnector(
```
probably the style followed in `worker.setTargetState` is preferable

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -1221,31 +1297,56 @@ public Void call() throws Exception {
             public Void call() throws Exception {
                 try {
-                    startConnector(connectorName);
+                    startConnector(
```
this can also be written in a less verbose way (see also comment above)

C0urante replied:
Not sure what you're referring to with `Worker::setTargetState`, could you elaborate?

kkonstantine replied:
I'm referring to how the lambda was written there as the second argument. I'd be nice to keep a consistent style at least in the changes included in a single commit.

C0urante replied:
Ahh, I see. I don't think that's as easy with this part since the `if` statement requires a new scope and the braces for that scope come with their own requirements about being on new lines.

kkonstantine replied:
I suggested what you've written in `StandaloneHerder` for the same method. Anyway that's minor.

kkonstantine reviewed (commented):
I completed a pass in the non testing code and left a few minor comments. 
I'll review the tests next.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -240,11 +243,19 @@ public void testJoinAssignment() throws Exception {
+        member.wakeup();
         EasyMock.expect(worker.isRunning(CONN1)).andReturn(true);
-
+        PowerMock.expectLastCall();
```
this seems to refer to `member.wakeup`. Best to follow that method.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -269,9 +280,17 @@ public void testRebalance() throws Exception {
+                return true;
+            }
+        });
+        member.wakeup();
```
This call is missing an `expectLastCall`. Here an in a couple other places.

C0urante committed:
KAFKA-9374: Address review comments

C0urante commented:
Thanks for the review @kkonstantine; I've pushed a new commit that tries to address your comments about whitespace and synchronization style. Ready for the testing comments whenever.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -582,86 +658,41 @@ public void testCreateConnectorFailedBasicValidation() throws Exception {
-        time.sleep(1000L);
-        assertStatistics(3, 1, 100, 1000L);
-
-        PowerMock.verifyAll();
```
don't we need to verify the expectations ?

C0urante replied:
The existing `testCreateConnectorFailedBasicValidation` and `testCreateConnectorFailedCustomValidation` tests were refactored into a single `testCreateConnectorFailedValidation` method, which does make the call to `PowerMock.verifyAll()`.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java:
```diff
@@ -676,63 +707,38 @@ public void testCreateConnectorFailedCustomValidation() throws Exception {
-        time.sleep(1000L);
-        assertStatistics(3, 1, 100, 1000L);
-
-        PowerMock.verifyAll();
```
same question. Not sure if there are not mocks used here.

C0urante replied:
Yeah, the only mock here is `connectorMock`, and no expectations are set on it. It's just used to trigger the `if (connector instanceof SinkConnector)` brach in `DistributedHerder::validateBasicConnectorConfig`.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java:
```diff
@@ -0,0 +1,488 @@
+    }
+
+    @Test
+    @Ignore("Interaction with this connector method has not been made asynchronous yet and this test will fail")
```
I think it's not great to introduce tests that will be ignored by default. As we wouldn't add unused code, maybe we want to skip adding these test cases for now.

C0urante replied:
Fair, I'll take them out. Will file a follow-up jira ticket to note the gaps here in case someone runs into problems with the yet-unaddressed connector methods in the future.

C0urante replied:
Filed https://issues.apache.org/jira/browse/KAFKA-10142 to note the gaps in this fix.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java:
```diff
@@ -0,0 +1,488 @@
+
+    private void createNormalConnector() {
+        Map<String, String> props = new HashMap<>();
+        props.put(CONNECTOR_CLASS_CONFIG, MonitorableSinkConnector.class.getName());
```
Should we create source connectors that actually produce data? Doesn't seem to be any flow of records to this connector.

C0urante replied:
Yeah, I think we can do that.

kkonstantine commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java:
```diff
@@ -0,0 +1,488 @@
+        waitForConnectorStart(NORMAL_CONNECTOR_NAME);
+    }
+
+    // TODO: Consider patching plugin scanning to handle connectors that block in their version methods
```
Code might not be the best way to document this, and it might get stale with time.

C0urante replied:
I'd like to leave some note in here about why `Connector::version` is a special case to avoid possible headaches in the future if/when someone tries to address it. I'll make it as minimal as possible and won't leave in any commented-out code; if you still think that's unwise I'll remove it, though.

kkonstantine reviewed (commented):
Completed a review of the tests too. 
Overall in the right direction, my comments are just related to specific details.

C0urante committed:
KAFKA-9374: Address review comments on tests

C0urante commented:
Thanks @kkonstantine; I've adjusted the integration test to verify that the well-behaved connector is not only instantiated by the framework, but also able to produce source records with its tasks. I've also added the calls to `PowerMock.expectLastCall()` for the member wakeups; good catch.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerder.java:
```diff
@@ -191,32 +189,59 @@ public synchronized void putConnectorConfig(String connName,
-                callback.onCompletion(new ConnectException("Failed to start connector: " + connName), null);
-                return;
-            }
+            // startConnector(connName, onStart);
```
nit: leftover?

C0urante replied:
Thanks, nice catch!

kkonstantine reviewed (approved):
Finished another pass and things look good. My minor comments/replies didn't add up to more than two. 

LGTM!

C0urante committed:
KAFKA-9374: Remove commented-out code

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -1221,31 +1284,56 @@ public Void call() throws Exception {
             public Void call() throws Exception {
                 try {
-                    startConnector(connectorName);
+                    startConnector(
+                            connectorName,
+                            (error, result) -> {
+                                if (error != null) {
+                                    log.error("Failed to start connector '" + connectorName + "'", error);
+                                }
+                            }
+                    );
```
```suggestion
                    startConnector(connectorName, (error, result) -> {
                        if (error != null) {
                            log.error("Failed to start connector '" + connectorName + "'", error);
                        }
                    });
```

C0urante committed:
KAFKA-9374: Style improvement

kkonstantine commented:
retest this please

kkonstantine commented:
ok to test

C0urante committed:
KAFKA-9374: Further style improvements

kkonstantine commented:
retest this please

C0urante commented:
Fixed the merge conflicts; should be safe to do a new test run now.

C0urante committed:
KAFKA-9374: Fix race condition in failing unit test

C0urante commented:
I believe I've identified the cause of the failure in https://builds.apache.org/job/kafka-pr-jdk14-scala2.13/1048. I've pushed a new commit to fix the logic; it passed tests locally, hopefully Jenkins will be happy with it.

kkonstantine commented:
retest this please

kkonstantine commented:
ok to test

kkonstantine commented:
retest this please

kkonstantine commented:
The underlying jenkins jobs of the latest builds were as follows: 
[JDK8](https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/2900/#showFailuresLink): 3 unrelated failures 
[JDK11](https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/6912/#showFailuresLink): 3 unrelated failures 
[JDK14](https://builds.apache.org/job/kafka-pr-jdk14-scala2.13/1063/testReport/): 2 unrelated failures 

(`kafka.admin.ReassignPartitionsUnitTest.testModifyBrokerThrottles` failed consistently with an NPE)

This changes here LGTM now. 
Thanks @C0urante for undertaking this useful task!
